substrate/client/rpc-spec-v2/src/chain_head/chain_head_storage.rs

```diff
@@ -18,192 +18,106 @@
-		&mut self,
-		mut block_guard: BlockGuard<Block, BE>,
-		hash: Block::Hash,
-		child_key: Option<ChildInfo>,
-	) {
-		let sender = block_guard.response_sender();
-		let operation = block_guard.operation();
-
-		while let Some(query) = self.iter_operations.pop_front() {
-			if operation.was_stopped() {
-				return
-			}
-
-			let result = self.client.query_iter_pagination(
-				query,
-				hash,
-				child_key.as_ref(),
-				self.operation_max_storage_items,
-			);
-			let (events, maybe_next_query) = match result {
-				QueryIterResult::Ok(result) => result,
-				QueryIterResult::Err(error) => {
-					send_error::<Block>(&sender, operation.operation_id(), error.to_string());
-					return
-				},
-			};
-
-			if !events.is_empty() {
-				// Send back the results of the iteration produced so far.
-				let _ = sender.unbounded_send(FollowEvent::<Block::Hash>::OperationStorageItems(
-					OperationStorageItems { operation_id: operation.operation_id(), items: events },
-				));
-			}
-
-			if let Some(next_query) = maybe_next_query {
-				let _ =
-					sender.unbounded_send(FollowEvent::<Block::Hash>::OperationWaitingForContinue(
-						OperationId { operation_id: operation.operation_id() },
-					));
-
-				// The operation might be continued or cancelled only after the
-				// `OperationWaitingForContinue` is generated above.
-				operation.wait_for_continue().await;
-
-				// Give a chance for the other items to advance next time.
-				self.iter_operations.push_back(next_query);
-			}
-		}
-
-		if operation.was_stopped() {
-			return
-		}
-
-		let _ =
-			sender.unbounded_send(FollowEvent::<Block::Hash>::OperationStorageDone(OperationId {
-				operation_id: operation.operation_id(),
-			}));
-	}
-
 	/// Generate the block events for the `chainHead_storage` method.
 	pub async fn generate_events(
 		&mut self,
-		mut block_guard: BlockGuard<Block, BE>,
 		hash: Block::Hash,
 		items: Vec<StorageQuery<StorageKey>>,
 		child_key: Option<ChildInfo>,
-	) {
-		let sender = block_guard.response_sender();
-		let operation = block_guard.operation();
-
-		let mut storage_results = Vec::with_capacity(items.len());
-		for item in items {
-			match item.query_type {
-				StorageQueryType::Value => {
-					match self.client.query_value(hash, &item.key, child_key.as_ref()) {
-						Ok(Some(value)) => storage_results.push(value),
-						Ok(None) => continue,
-						Err(error) => {
-							send_error::<Block>(&sender, operation.operation_id(), error);
-							return
-						},
-					}
-				},
-				StorageQueryType::Hash =>
-					match self.client.query_hash(hash, &item.key, child_key.as_ref()) {
-						Ok(Some(value)) => storage_results.push(value),
-						Ok(None) => continue,
-						Err(error) => {
-							send_error::<Block>(&sender, operation.operation_id(), error);
-							return
-						},
+		tx: mpsc::Sender<QueryResult>,
+	) -> Result<(), tokio::task::JoinError> {
+		let this = self.clone();
+
+		tokio::task::spawn_blocking(move || {
```

Comment:
Nice! We are doing a `spawn_blocking` here instead of in `chain_head_unstable_storage`. Looks good to me! Thanks for simplifying this module!